src/ortho.rs

```diff
@@ -1,10 +1,124 @@
+    } else {
+        let logical_coords = get_logical_coordinates(dimensions);
+        let current_shell: u16 = logical_coords[storage_length].iter().sum();
+        
+        let forbidden_indices: Vec<usize> = (0..storage_length)
+            .filter(|&index| {
+                if index < logical_coords.len() {
+                    let coords = &logical_coords[index];
+                    let shell: u16 = coords.iter().sum();
+                    shell == current_shell
+                } else {
+                    false
+                }
+            })
+            .collect();
+        
+        cache_guard.insert(cache_key, forbidden_indices.clone());
+        forbidden_indices
+    }
+}
+/// Get cached required coordinate indices or compute and cache them
+fn get_required_coordinate_indices(dimensions: &[u16], current_logical: &[u16]) -> Vec<Vec<usize>> {
+    let cache = REQUIRED_INDICES_CACHE.get_or_init(|| Mutex::new(HashMap::new()));
+    let cache_key = (dimensions.to_vec(), current_logical.to_vec());
+    let mut cache_guard = cache.lock().unwrap();
+    
+    if let Some(indices) = cache_guard.get(&cache_key) {
+        indices.clone()
+    } else {
+        let logical_coords = get_logical_coordinates(dimensions);
+        
+        // Stage 1: Generate the list of list of indices for coordinates satisfying the property 
+        // that each list of coordinates traverses one axis from the edge to the given position (not inclusive)
+        let required_indices: Vec<Vec<usize>> = (0..dimensions.len())
+            .map(|axis| {
+                (0..current_logical[axis])
+                    .filter_map(|coord_value| {
+                        let mut coords = current_logical.to_vec();
+                        coords[axis] = coord_value;
+                        // Find the index of these coordinates in our logical coordinate system
+                        logical_coords.iter().position(|c| c == &coords)
+                    })
+                    .collect()
+            })
+            .collect();
+        
+        cache_guard.insert(cache_key, required_indices.clone());
+        required_indices
+    }
+}
+
+/// Get required values by computing coordinate indices and mapping to storage
+fn get_required_values(dimensions: &[u16], storage: &[u16]) -> Vec<Vec<u16>> {
+    let logical_coords = get_logical_coordinates(dimensions);
+    let current_logical = logical_coords[storage.len()].clone();
+    
+    // Get cached coordinate indices
+    let required_indices = get_required_coordinate_indices(dimensions, &current_logical);
+    
+    // Stage 2: Turn those indices into values contained by the storage 
+    // by looking them up directly
+    let required: Vec<Vec<u16>> = required_indices.into_iter()
+        .map(|index_list| {
+            index_list.into_iter()
+                .filter_map(|index| {
+                    // Look up the stored value at that index
+                    if index < storage.len() {
```

Comment:
This has logical coordinate handling and so should be under the cache. Only the lookup should be out here. It could literally be a map call to a lookup.